docs/Compiler-Hardening-Guides/Compiler-Annotations-for-C-and-C++.md

```diff
@@ -0,0 +1,42 @@
+# Compiler Annotations for C and C++
+
+Compile time security analysis and runtime mitigation implemented in compilers both depend on the compiler being able to see the flow of data between different points in a program, across functions and modules. This is quite a challenge in C and C++ because both languages allow passing around opaque references, thus losing information about objects.  To work around this problem, both GCC and Clang implement attributes to annotate source code, especially functions and data structures, to allow them to do better analysis of source code.  These annotations are not only beneficial for security, but they also help the compilers make better optimization decisions, often resulting in better code.
+
+## Clang and GCC Attributes at a glance
+
+Both GCC and Clang recognize the `__attribute__` keyword to annotate source code.  Both compilers also provide a `__has_attribute()` macro function that returns 1 if the attribute name passed to it is supported and 0 otherwise.  For example `__has_attribute(malloc)` would return 1 in the latest GCC and Clang.  The full syntax description for the `__attribute__` keyword is available in the GCC documentation[^1].  Since all of the attributes recommended in this document are function attributes, here's a quick primer on how they are used.
```

Comment:
I support the general idea of encouraging annotation that helps improve safety and security.  However, I have concerns about the syntax recommendations.  Where possible, we should encourage standard notation.  For instance, `[[fallthroug]]` is a standard C++ notation, now also available in C23.  I think a recommendation along the lines of "use these standard notations where available, or fall back these non-standard notations" would be a better recommendation.

Reply:
Ack, I wanted to put this work out for early feedback. I can modify examples (and indeed, all of the text) to use the C/C++ standard notation.

Reply:
One challenge is that code often needs to be compilable with many existing and older compilers. So syntax from new specs often takes a while before people can use it in code. We should mention a spec where it exists, but also note an alternative when that option isn't available to a significant number of cases.

Reply:
> So syntax from new specs often takes a while before people can use it in code. We should mention a spec where it exists, but also note an alternative when that option isn't available to a significant number of cases.

I do agree; it is worth noting that the standard notations have been available in at least the 5 major releases or GCC (which is more than supported/maintained LTS releases), and many more in GCC.   In our guidelines, I think we should strive to guide users to the standard notations first, where available, and fallback to non-standard notations only when no alternatives exist.

Reply:
If I understand you correctly I agree 

Let's document and recommend and prefer standard forms. Let's also document the alternatives in the unfortunate cases where that won't work for some reason.